File: clients/python/examples/people.py
"""Use the same builders with either Corium peer topology."""

from __future__ import annotations

import asyncio
import os

from corium import (
    EntityMap,
    LocalPeer,
    Peer,
    Pull,
    Query,
    RemotePeer,
    TxBuilder,
    attr,
    data,
    gte,
    lookup,
    tempid,
)


async def load_and_query(peer: Peer) -> None:
    await peer.transact(
        TxBuilder()
        .entity(
            EntityMap.with_id(tempid("ada"))
            .set("person/name", "Ada")
            .set("person/age", 36)
        )
        .build()
    )

    adults = (
        Query.find_collection("?name")
        .in_scalar("?minimum")
        .where(data("?entity", attr("person/name"), "?name"))
        .where("?entity", ":person/age", "?age")
        .where(gte("?age", "?minimum"))
    )
    db = await peer.sync()
    print(await db.query(adults, 18))
    print(
        await db.pull(
            Pull().db_id().attr("person/name").attr("person/age"),
            lookup("person/name", "Ada"),
        )
    )


async def main() -> None:
    endpoint = os.environ.get("CORIUM_ENDPOINT", "http://127.0.0.1:4334")
    database = os.environ.get("CORIUM_DATABASE", "people")
    if os.environ.get("CORIUM_REMOTE"):
        peer: Peer = await RemotePeer.connect(endpoint, database=database)
    else:
        peer = await LocalPeer.connect(endpoint, database=database)
    async with peer:
        await load_and_query(peer)


if __name__ == "__main__":
    asyncio.run(main())

File: clients/java/examples/People.java
import io.corium.client.EntityMap;
import io.corium.client.Query;
import io.corium.client.RemotePeer;
import io.corium.client.TxBuilder;

public final class People {
    public static void main(String[] args) {
        String endpoint = args.length > 0 ? args[0] : "http://127.0.0.1:4336";
        try (RemotePeer peer = RemotePeer.builder(endpoint, "people").build()) {
            var report = peer.transact(new TxBuilder().entity(
                    EntityMap.withId("ada")
                            .set("person/name", "Ada")
                            .set("person/age", 36))).join();
            var adults = Query.findCollection("?name")
                    .inScalar("?minimum")
                    .where("?entity", ":person/name", "?name")
                    .where("?entity", ":person/age", "?age")
                    .where(Query.gte("?age", "?minimum"));
            System.out.println(report.dbAfter().query(adults, 18).join().value());
        }
    }
}
